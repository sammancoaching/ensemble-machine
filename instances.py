#!/usr/bin/env python3
import datetime
import json
from dataclasses import dataclass
from pathlib import Path

import boto3
import click

from summon import read_regions_config

@dataclass
class RunningInstance:
    ip_address: str
    state: str
    coach: str
    url: str
    creation_date: str


@click.command()
@click.option(
    "--region-name",
    help=f"the aws region name (default eu-north-1): {', '.join(read_regions_config().keys())}",
    default="eu-north-1"
)
@click.option(
    "--aws-profile",
    default=None,
    help="the aws profile, if you dont use the default"
)
@click.option(
    "--coach",
    help="only show instances owned by this person",
)
def main(region_name, aws_profile, coach=None):
    instances = all_instances(region_name, aws_profile)
    if coach:
        instances = [instance for instance in instances if instance.coach.strip() == coach]
    print('\n'.join(print_instances(instances)))


def all_instances(region_name, aws_profile):
    response = get_sammancoach_machines(region_name, aws_profile)
    instances = instances_from_response(response)
    return instances


def print_instances(instances):
    rows = []
    for instance in instances:
        rows.append(print_instance(instance))
    return rows


def get_sammancoach_machines(region_name, aws_profile):
    session = boto3.Session(profile_name=aws_profile, region_name=region_name)
    client = session.client('ec2')
    custom_filter = [{
        'Name': 'tag:SammanCoach',
        'Values': ['*']}]
    response = client.describe_instances(Filters=custom_filter)
    return response


def instances_from_response(obj):
    machines = []
    for reservation in obj['Reservations']:
        for instance in reservation['Instances']:
            tags = instance['Tags']
            public_ip = instance.get("PublicIpAddress", "not available")
            attach_time = try_find_attach_time(instance)
            attach_time_short = attach_time.partition('T')[0] if attach_time else '-'
            machine_data = {
                "State": instance["State"]["Name"],
                "IP": public_ip,
                "AttachTime": attach_time_short,
            }
            for machine in tags:
                key = machine['Key']
                value = machine['Value']
                machine_data[key] = value
            machines.append(machine_data)
    instances = []
    for machine in machines:
        # TODO: find better way to filter out projector machines!
        if 'codekata.proagile.link' in machine['Name']:
            instance = RunningInstance(ip_address=f"{machine['IP']:16}",
                                       state=f"{machine['State']:12}",
                                       coach=f"{machine['SammanCoach']:12}",
                                       url=f"https://{machine['Name']:42}",
                                       creation_date=f"{machine['AttachTime']}")
            instances.append(instance)
    return instances


def print_instance(instance: RunningInstance):
    return f"{instance.ip_address} {instance.state} {instance.coach} {instance.url} {instance.creation_date}"


def try_find_attach_time(instance):
    try:
        time_value = instance["BlockDeviceMappings"][0]["Ebs"]["AttachTime"]
        if isinstance(time_value, datetime.datetime):
            return time_value.isoformat()
        else:
            return time_value
    except:
        return None


if __name__ == '__main__':
    import logging

    logging.basicConfig(level=logging.INFO)
    main()
