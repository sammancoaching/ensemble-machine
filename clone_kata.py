#!/usr/bin/env python3

import click
import os
from fabric import Connection
from dataclasses import dataclass
import csv

from instances import all_instances
from summon import read_regions_config, read_aws_defaults


@dataclass
class KataMachine:
    region_name: str
    host_ip: str
    url: str


def read_classroom_file(file, running_instances):
    result = []
    reader = csv.DictReader(file)
    for row in reader:
        url = row["url"]
        region_name = row["region"]
        matching_instances = [machine for machine in running_instances if machine.url.strip() == url.strip()]
        if matching_instances:
            instance = matching_instances[0]
            machine = KataMachine(url=url, region_name=region_name, host_ip=instance.ip_address)
            result.append(machine)
        
    return result


def clone_kata_commandline(kata):
    return f"""
                sudo -u typist -- sh -c '\
                   cd /home/typist
                   mkdir -p katas;\
                   cd katas;\
                   git clone {kata}
                '\
              """


@click.command()
@click.option(
    "--kata",
    help="the kata to clone",
    default="https://github.com/emilybache/starter.git",
    prompt=f"what kata repo?"
)
@click.option(
    "--region-name",
    help=f"the aws region name (default eu-north-1): {', '.join(read_regions_config().keys())}",
    default="eu-north-1"
)
@click.option(
    "--aws-profile",
    default=None,
    help="the aws profile, if you don't use the default"
)
@click.option(
    "--host_ip",
    help="the ip address",
)
@click.option(
    "--coach",
    help="add the kata to all running instances owned by this person",
)
@click.option(
    "--classroom",
    help="add the kata to all running instances in this classroom file",
)
def clone_kata(kata, region_name, aws_profile, host_ip, coach, classroom):
    commandline = clone_kata_commandline(kata)
    aws_defaults = read_aws_defaults(profile_name=aws_profile)
    url_stem = aws_defaults["url_stem"]
    machines = determine_machines_to_update(aws_profile, classroom, coach, host_ip, region_name, url_stem)
    logging.getLogger().info(f"will clone kata to machines: {[m.url for m in machines]}")
    run_commandline_on_machines(commandline, machines, aws_profile)


def run_commandline_on_machines(commandline, machines, aws_profile):
    for m in machines:
        try:
            print(f"will run remote commands on {m}")
            c = connect_to_machine(m, aws_profile)
            c.run(commandline)
        except Exception as e:
            print("unexpected problem running remote commands on machine ", m, e)


def connect_to_machine(machine, aws_profile):
    key_name = read_regions_config(profile_name=aws_profile)[machine.region_name]["key_name"]
    key_file = os.path.expanduser(f"~/.ssh/{key_name}.pem")
    c = Connection(host=machine.host_ip.strip(), user='ubuntu', connect_kwargs={
        "key_filename": key_file,
    }, )
    return c


def determine_machines_to_update(aws_profile, classroom, coach, host_ip, region_name, url_stem):
    running_instances = [machine for machine in all_instances(region_name, aws_profile) if
                         machine.state.strip() == "running"]
    if not running_instances:
        logging.getLogger().error(f"No running instances found in region {region_name}")
        return []
    if host_ip:
        instance = [machine for machine in running_instances if host_ip in machine.ip_address]
        if not instance:
            print(f"ERROR: host ip {host_ip} is not running")
            machines = []
        else:
            machines = [KataMachine(host_ip=host_ip, region_name=region_name, url=instance[0].url)]
    elif coach:
        running_instances = [machine for machine in running_instances if machine.coach.strip() == coach]
        machines = [KataMachine(host_ip=machine.ip_address, region_name=region_name, url=machine.url) for machine in
                    running_instances]
    elif classroom:
        with open(classroom, 'r', newline="", encoding="utf-8") as f:
            machines = read_classroom_file(f, running_instances)
    else:
        machines = [KataMachine(host_ip=machine.ip_address, region_name=region_name, url=machine.url) for machine in
                    running_instances if url_stem in machine.url]
    return machines


if __name__ == "__main__":
    import logging

    logging.basicConfig(level=logging.INFO)
    clone_kata()
