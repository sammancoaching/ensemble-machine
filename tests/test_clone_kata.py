from io import StringIO

from approvaltests import verify_all

from clone_kata import read_classroom_file, KataMachine
from instances import RunningInstance


def test_read_classroom_file():
	classroom_file = """\
room,region,id,url,team,comments
1,ca-central-1,i-0bd2b004ba4473c6e,https://c7f3aa50-1-idea.codekata.proagile.link,,
2,ca-central-1,i-07f1241bff8f0d1ee,https://c7f3aa50-2-idea.codekata.proagile.link,,
"""
	f = StringIO(classroom_file)
	result = read_classroom_file(f, running_instances=[
		RunningInstance("18.157.73.25", "running", "emily", "https://c7f3aa50-1-idea.codekata.proagile.link", "2022-01-31", region_name="eu-north-1"),
		RunningInstance("18.157.73.26", "running", "emily", "https://c7f3aa50-2-idea.codekata.proagile.link", "2022-01-31", region_name="eu-north-1"),
	])

	verify_all("machines", result)