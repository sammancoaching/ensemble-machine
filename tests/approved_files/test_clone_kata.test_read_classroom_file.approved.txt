machines

0) KataMachine(region_name='ca-central-1', host_ip='18.157.73.25', url='https://c7f3aa50-1-idea.codekata.proagile.link')
1) KataMachine(region_name='ca-central-1', host_ip='18.157.73.26', url='https://c7f3aa50-2-idea.codekata.proagile.link')
